### ISSUE TITLE
port to python3

### ISSUE THREAD

**Issue by lynxlynxlynx** [gemrb/gemrb#101]
python2 will get EOL-ed in 2020, so we need to move forward, to python3. Even exotic platforms have a port by now, so we can consider dropping v2 support if maintaining compatibility with both would be a hassle.

Pointless to start work before #95 is done though, since that already includes substantial changes to the GUIScripts plugin.

**Comment by bradallred** [comment]
the scripts themselves are probably in good order save for some minor issues. as You point out the GUIScript will be a tedious pain to update. At a minimum all the string conversion functions need to change.

**Comment by lynxlynxlynx** [comment]
I think 2to3 will take care of a large chunk for us, so I'm not that worried. Save for the actual two modules and the metaclass stuff, we don't do anything advanced.

**Comment by bradallred** [comment]
The masochist in me wants to suggest just using pybind now that we have C++11. Pybind is really nice and easy, but we have the hitch of Window/Control references to deal with. The main issue being they need to be `delete`able from C++ side without blowing up the Python side. We would have to ditch our ScriptingRef layer and instead use `std::shared_ptr` for the `WindowManager::windows` and the `Window::controls` and serve up `std::weak_ptr` objects to the GUIScripts... assuming that even works (I have no reason to believe it wouldn't from my experience with pybind). We would have to proof-of-concept that first.

I'm sure it would be a nicer interface with much less code, but I'm less convinced that the effort is worth it unless the effort to update what we do now (GUIScript.cpp) is also high.

The question is what are the removed/deprecated cython APIs we are using that are not easily updated. At a minimum we will have a bunch of cstring fixing to do.

**Comment by MarcelHB** [comment]
pybind sounds like the dream of scripting C++-things in Python.

Anyway, is there something at the horizon that makes GUIScript.cpp subject to larger changes (after subviews)?

**Comment by lynxlynxlynx** [comment]
I don't think so, but let's keep focused and deal with this later.

**Comment by bradallred** [comment]
> pybind sounds like the dream of scripting C++-things in Python.

I truly can't see how anything could be better :) I use it all the time and it makes (most) things trivial.

> Anyway, is there something at the horizon that makes GUIScript.cpp subject to larger changes (after subviews)?

This is the closest thing I know of.

> I don't think so, but let's keep focused and deal with this later.

Of course. This is tagged "post-subviews" for a reason.

**Comment by bradallred** [comment]
this is going to be more work than anticipated. something is wrong with the meta class and I have no idea what.

**Comment by bradallred** [comment]
oh good, it looks like the way you [define the meta class](https://portingguide.readthedocs.io/en/latest/classes.html#metaclasses) just changed

**Comment by lynxlynxlynx** [comment]
2to3 also suggests some changes to the python side in MetaClasses.py

**Comment by bradallred** [comment]
we can steal this from six, which works fine

https://github.com/benjaminp/six/blob/1a4f325d55b10541ea94b889cda01e6281ba1689/six.py#L880

**Comment by lynxlynxlynx** [comment]
Looks like the rest of 2to3 changes are few and far between, with some false positives in print().

**Comment by bradallred** [comment]
ok, that appears to bed all the heavy lifting done.

from here we should just be able to run 2to3 to fix up some syntax... maybe add some try+except for `dict.iteritems` like was done with the metaclass

`GemRB_Button_SetHotKey` needs to be fixed to run anything, but I hacked it to return `PyNone` just so I could make sure BG2 start screens could load. loading a game still breaks due to some syntax errors and the like.

**Comment by lynxlynxlynx** [comment]
Nice! 2to3 has almost no changes any more, with a bunch of useless print suggestions (and a bunch correct ones). Many are compatible with both versions, so I can whittle the diff down further.

**Comment by bradallred** [comment]
We can now build and "run" with Python 3. There are tons of errors I haven't looked into when loading a game. Hopefully easy hurdles.

**Comment by bradallred** [comment]
we can now load into a game without errors (only tested a single BG2 save tho)

the portrait icons are busted... several windows still produce errors... journal entries have floating point numbers... the rabbit hole is deep

**Comment by lynxlynxlynx** [comment]
Are we basically still struggling only with the unicode changes? I can go through all our functions to see which ones still need work.

**Comment by lynxlynxlynx** [comment]
Audio works for me and I've fixed a few minor issues. Changing cmake to look for py3 requires just one line, so the switch is trivial.

**Comment by bradallred** [comment]
the audio issue is specifically with BG2 start music and is defiantly only a problem when switching to python3

**Comment by lynxlynxlynx** [comment]
Ah, I see what you mean now. It's the soa/tob music that doesn't start, once you choose, the first window has it.

**Comment by lynxlynxlynx** [comment]
I tried all the games, cg, main windows, dialog ... no new issue. :) Even bg1 chargen with those __next__ changes, which are not required. So 2to3 can be laid to rest and we've achieved py3 compat without needing to sacrifice py2. I also reviewed all the divisions, so floats hopefully are not a problem anywhere any more.

What we already knew about:
- [ ] status icons both on portrait win and in guirec description
- [ ] no music in soa/tob start2 menu
